src/index.tsx

```diff
@@ -62,67 +68,102 @@ export declare type SnapyrInAppMessage = {
   content: SnapyrInAppContent;
 };
 
-export const SnapyrEmitter = new NativeEventEmitter(SnapyrRnSdk);
-// Client code can register listeners (callbacks) on SDK events; use a map to limit to 1 listener per event type
-const _eventListeners = new Map<string, EmitterSubscription>();
+export declare type SnapyrPushNotificationPayload = {
+  notificationId: number;
+  titleText: string;
+  contentText: string;
+  subtitleText: string | null;
+  deepLinkUrl: string | null;
+  imageUrl: string | null;
+};
 
-export function onSnapyrDidRegister(callback: (token: string) => void): void {
-  const listener = SnapyrEmitter.addListener(
-    SNAPYR_LISTENER_REGISTER,
-    (token) => callback(token)
-  );
-  // Remove/unsubscribe previous listener, if any
-  _eventListeners.get(SNAPYR_LISTENER_REGISTER)?.remove();
-  _eventListeners.set(SNAPYR_LISTENER_REGISTER, listener);
-}
+// Same data for now but will likely change, e.g. to include action button ID or cancellation data, so establishing separate type
+export declare type SnapyrPushNotificationResponsePayload =
+  SnapyrPushNotificationPayload;
 
-export function onSnapyrDidReceiveNotification(
-  callback: (notification: any) => void
-): void {
-  const listener = SnapyrEmitter.addListener(
-    SNAPYR_LISTENER_NOTIFICATION,
-    (notification) => {
-      callback(notification);
-    }
-  );
-  // Remove/unsubscribe previous listener, if any
-  _eventListeners.get(SNAPYR_LISTENER_NOTIFICATION)?.remove();
-  _eventListeners.set(SNAPYR_LISTENER_NOTIFICATION, listener);
-}
+type SnapyrEventCallback<T> = (eventPayload: T) => void;
 
-export function onSnapyrInAppMessage(
-  callback: (message: SnapyrInAppMessage) => void
-): void {
-  const listener = SnapyrEmitter.addListener(
-    SNAPYR_LISTENER_INAPP_MESSAGE,
-    (message: SnapyrInAppMessage) => {
-      callback(message);
+export const SnapyrEmitter = new NativeEventEmitter(SnapyrRnSdk);
+// Client code can register listeners (callbacks) on SDK events; use a map to limit to 1 listener per event type
+const _eventCallbacks = new Map<string, SnapyrEventCallback<any>>();
+// If we receive an event before client has registered a callback for it, store the payload so it can optionally be fired once a callback is registered. Only stores one payload (the last one to be received) for each event at this time.
+const _pendingEvents = new Map<string, any>();
+
+/**
+ * Generate an event listener registration callback for the given (native) event name, and start listening for native events.
+ *
+ * This will catch and store any event fired from the native side even if no user callback has been registered yet, allowing for a "replay" of the latest event at callback registration time.
+ */
+function setupEventListener<T>(
```

Comment:
This is a function generator to reduce boilerplate for native event listener setup. I had more events/listeners partway through dev than I ended up with, making this seem more necessary, but hopefully it'll be useful going forward. Native events going to React Native always work the same way - an event name, and a callback that is passed a single object (which can contain arbitrary serializable data).

This also implements the newer logic to have the SDK immediately register listeners and stash events that arrive prior to the first user-provided callback registration.